### DISCUSSION TITLE
Vibrance

### DISCUSSION THREAD

**Discussion by Solido** [libvips/libvips#4039] [Ideas]
I'm looking to integrate Vibrance aka Smart Saturation.
Subjects is not much discussed while the feature is of much importance.
Thank you for you suggestions on the matter!

**Comment by jcupitt** [comment]
I had a quick google and found:

> Saturation impacts all color intensity equally, which is why it's so easy to go overboard so quickly. Vibrance, on the other hand, only increases the intensity of the less saturated colors in an image while simultaneously trying to avoid skin tones and prevent the gaudy posterization that happens when you crank your saturation up to the max.

Sure, it sounds like a simple thing to implement. A detailed description, sample images and expected results would be very helpful.

**Reply by Solido** [reply]
Hi John! I've started a discussion in Image: https://github.com/elixir-image/image/discussions/147#discussioncomment-10041187 on this matter and others related subjects.

**Comment by Solido** [comment]
It seems to exist many implementations. I'm myself not sure but please consult the previous link to have an idea.

The goal is too enhance colors of an image without distorting too much the tonal equilibrium like saturation would do.

0
![bird-0](https://github.com/user-attachments/assets/d4a4a7f3-8bc1-4e1f-bd62-8eb105ea42cf)

-50%
![bird-minus-50](https://github.com/user-attachments/assets/5e4c7d5c-ba08-46f1-a831-f24bbba7f702)

-100%
![bird-minus-100](https://github.com/user-attachments/assets/b7f9d785-f8b8-4a16-9aba-fb7e00e69fdf)

+50%
![bird-plus-50](https://github.com/user-attachments/assets/f726ec0c-1b35-4a87-8f0e-58875b32bea5)

+100%
![bird-pus-100](https://github.com/user-attachments/assets/572428af-75a7-4da5-bd89-7afd09cc8cfb)

**Reply by jcupitt** [reply]
It sounds like the photoshop one has some extra stuff to avoid boosting skin tones. Do you have any samples of this in action?

**Reply by Solido** [reply]
From a pure photography perspective, vibrance is much more extended that skin tones. We may even use other technics as we want to affect only a specific color range.

Right now I use Affinity Photo, which is free for 6 months as trial. You may enjoy what the tool offer and test the effect to decipher a probable implementation.

Would be great in Nip2 too!

**Reply by jcupitt** [reply]
Sure, but from the description above ^^ it sounds like, as well as avoiding oversaturating already saturated colours, photoshop has some extra logic to prevent skin tones (not saturated) from blowing out.

Perhaps we can consider that separately.

**Reply by Solido** [reply]
Skin is such a subject! 

it's a whole know-how from tone, texture (no too much details), specular, etc
The model will juge you as an artist only by this matter :)
This would be the cherry on top but maybe last feature to add?

Having some controls over Vibrance, Clarify, Structure,...
would cover a LOT from portrait to landscape and street.

Since I'm new here, I do not want to ask for too much features
nor I'm sure where it belongs (vips, vix, image) but I would gladly
help with a roadmap, testing, reviews.

Can you help clarify the scope of lipvips? 
should it stay generic or can I ask for subtile/esthetic features
than we may found in photo editor?

**Reply by jcupitt** [reply]
I think libvips should be very generic. Stuff like that should be in the applications that use it, not in the core library. 

But ... they should be simple to implement (just a few lines of code). Making a prototype in eg. python is a good way to see if the library needs a new feature.

**Comment by jcupitt** [comment]
I made a tiny nip2 workspace with a prototype:

![image](https://github.com/user-attachments/assets/8489ff5b-fc84-4b0f-b934-2f6f95ce2473)

That's using the RHS of the logistic function to transform C in LCh.

Sample image:

![bird](https://github.com/user-attachments/assets/c89a678d-aa8e-4ff8-8832-56c122cf9611)

With +0.5:

![bird2](https://github.com/user-attachments/assets/e2a81e36-543d-49af-9cec-13ef30749ff5)

The range might need a bit of tuning.

**Reply by jcupitt** [reply]
Workspace here:

http://www.rollthepotato.net/~john/vibrance.ws

**Reply by Solido** [reply]
Thanks! I would like to use Nip2 but I failed my installation on mac.
No formula nor build. Not sure what I'm missing.

The goal of Vibrance is to _flatten_ the saturation or align chroma without going through the roof :D

I've got some nice result by applying a tone on HSV.
Why are you using LCh model? is there some benefits about colors compression?

Feels like I fall into a trap, so much cool stuffs to explore :D

ps: on your exemple, .5 should let the yellow and red intact, only background color should have moved up in chroma.

**Reply by jcupitt** [reply]
You have to go back a little way unfortunately (it has become very hard to build for macos), but this one works:

https://github.com/libvips/nip2/releases/download/v8.6.0-beta2/nip2-8.6.0-beta2.app.dmg

I don't like HSV much, LCh should be quite a bit better, hopefully.

Yes, we should tune it a bit -- I set it to leave C 100 alone but boost colours below that. Maybe 70 would be better.

**Reply by Solido** [reply]
Here's my tuning of @kipcole9 first impl.
I clamp left and right extension for preservation
then put a focus on shadow and lower degree to mids
resulting in an effective Vibrance.

We can select a default equilibrium between all those entries then
simply apply strength.

Not sure if the final formula belong to Vips as it will be a
strong 'parti pris'


``` elixir
Image.with_colorspace(image, :hsv, fn hsv ->
    toneOptions = [
      {:black_point, 10}, {:white_point, 90},
      {:shadows, 20}, {:shadow_point, 0.3}, 
      {:mid_points, 15}, {:mid_point, 0.5}, 
      {:highlights, 0}, {:highlight_point, 0.7}, 
    ]
    
    {:ok, mapped} = Image.apply_tone_curve(hsv[1], toneOptions)
    Image.join_bands([hsv[0], mapped, hsv[2]])
  end)
```

**Comment by kipcole9** [comment]
John, thank you as always. Understood on why LCh interpretation, makes sense (after yet more reading).

I can mentally convert the NIP2 workspace into code with the exception of `H1`. What libvips function does that correspond to?

> I set it to leave C 100 alone but boost colours below that. Maybe 70 would be better.

What coefficient drives that cutoff?

Thanks for taking me to school - again :-)

**Reply by jcupitt** [reply]
`H1` is just `vips_identity()`.

The 100 pivot point is in `G5` and `G7`, swap those 100s for 70s to try a lower value.

Drag the `H7` slider to try different amounts of vibrance.

LCh is a bit like HSV, but "scientific". LCh aims to be perceptually uniform, that is, any two points the same distance apart in LCh should be visibly different by the same amount to a human observer. 

This is important if you are transforming colours mathematically. An operation in LCh will transform all colours by approximately the same amount, but the same operation in HSV will boost some colours much more than others. 

LCh should look accurate and stable, HSV is a quick and ugly hack used in older image processing systems. It's only in libvips to make it easier to port very old code, you shouldn't use it in anything new.

**Comment by kipcole9** [comment]
Thanks John. Apologies but I have two last questions (related to my ignorance):

1. G7 `hist_map` maps to the function `maplut` if I understand correctly?
2. What is the function that implements `constant / image` math?   Normally I'd reach for `linear` but thats only appropriate for `image / constant`. I couldn't see an implementation in `rubyvips` or `pyvips` with this argument order.

**Reply by kipcole9** [reply]
Hmmm, I supposed I can take the constant value and create a single band image of the same dimensions as the image with that value and then just do the normal `image / image` division?

**Reply by jcupitt** [reply]
1. Yes, that's right.
2. Do reciprocal then linear, ie. $a / b = b a^{-1}$. You can use `pow` for the reciprocal.

**Reply by jcupitt** [reply]
pyvips does this here:

https://github.com/libvips/pyvips/blob/master/pyvips/vimage.py#L1524-L1530

ruby doesn't allow `number / object`, since arithmetic operators are really messages sent to the left-hand side, so it's not possible there, bizarrely. You'd need to extend the number class with a method for ruby-vips, which I've always balked at heh.

**Reply by jcupitt** [reply]
And C++ here (might be more obvious):

https://github.com/libvips/libvips/blob/master/cplusplus/VImage.cpp#L1062-L1066

**Reply by kipcole9** [reply]
Done, implemented in my lib - thanks again for all the help.

**Reply by Solido** [reply]
Thanks @jcupitt @kipcole9 it's a very enriching process trying to follow along!
